Looks like CI doesn't run any unit tests with the `snapshot` build tag.  We should probably do that.  With that said:

```
$ go test ./internal/pkg/ver/ -run TestCheckCompatibilityInternal -tags snapshot -v
```

```
=== RUN   TestCheckCompatibilityInternal
=== RUN   TestCheckCompatibilityInternal/empty_fleet_and_elasticsearch_version
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"","elasticsearch_version":"","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
    check_test.go:95: expected error
=== RUN   TestCheckCompatibilityInternal/empty_fleet_version
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"","elasticsearch_version":"8.0.0","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
    check_test.go:95: expected error
=== RUN   TestCheckCompatibilityInternal/empty_elasticsearch_version
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"7.13","elasticsearch_version":"","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
    check_test.go:95: expected error
=== RUN   TestCheckCompatibilityInternal/supported_elasticsearch_713-713
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"7.13.0","elasticsearch_version":"7.13.0","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
=== RUN   TestCheckCompatibilityInternal/supported_elasticsearch_7131-7132
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"7.13.2","elasticsearch_version":"7.13.1","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
=== RUN   TestCheckCompatibilityInternal/supported_elasticsearch_713-714
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"7.13.2","elasticsearch_version":"7.14.2","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
=== RUN   TestCheckCompatibilityInternal/supported_elasticsearch_715-800
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"7.15.2","elasticsearch_version":"8.0.0","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
=== RUN   TestCheckCompatibilityInternal/unsupported_elasticsearch_714-713
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"7.14.0","elasticsearch_version":"7.13.1","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
    check_test.go:95: expected error
=== RUN   TestCheckCompatibilityInternal/unsupported_elasticsearch_800-718
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"8.0.0","elasticsearch_version":"7.18.0","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
    check_test.go:95: expected error
=== RUN   TestCheckCompatibilityInternal/supported_elasticsearch_800a1
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"8.0.0-alpha1","elasticsearch_version":"8.0.0-alpha1","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
=== RUN   TestCheckCompatibilityInternal/supported_elasticsearch_715-800a1
    writer.go:27: es_check_snapshot.go:22: {"level":"info","fleet_version":"7.15.2","elasticsearch_version":"8.0.0-alpha1","message":"SNAPSHOT build skipping Elasticsearch compatibility check"}
--- FAIL: TestCheckCompatibilityInternal (0.00s)
    --- FAIL: TestCheckCompatibilityInternal/empty_fleet_and_elasticsearch_version (0.00s)
    --- FAIL: TestCheckCompatibilityInternal/empty_fleet_version (0.00s)
    --- FAIL: TestCheckCompatibilityInternal/empty_elasticsearch_version (0.00s)
    --- PASS: TestCheckCompatibilityInternal/supported_elasticsearch_713-713 (0.00s)
    --- PASS: TestCheckCompatibilityInternal/supported_elasticsearch_7131-7132 (0.00s)
    --- PASS: TestCheckCompatibilityInternal/supported_elasticsearch_713-714 (0.00s)
    --- PASS: TestCheckCompatibilityInternal/supported_elasticsearch_715-800 (0.00s)
    --- FAIL: TestCheckCompatibilityInternal/unsupported_elasticsearch_714-713 (0.00s)
    --- FAIL: TestCheckCompatibilityInternal/unsupported_elasticsearch_800-718 (0.00s)
    --- PASS: TestCheckCompatibilityInternal/supported_elasticsearch_800a1 (0.00s)
    --- PASS: TestCheckCompatibilityInternal/supported_elasticsearch_715-800a1 (0.00s)
FAIL
FAIL	github.com/elastic/fleet-server/v7/internal/pkg/ver	0.317s
FAIL
```

We'll want to update these test cases or pull them out into a separate file with the `!snapshot` build tag (or maybe leave them in the same file and add the build tag?) and also add unit tests for the snapshot implementation in this PR with a `snapshot` build tag.

--- a/internal/pkg/ver/es_check_snapshot_test.go
+++ b/internal/pkg/ver/es_check_snapshot_test.go
@@ -1,0 +1,66 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+//go:build snapshot
+
+package ver
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	testlog "github.com/elastic/fleet-server/v7/internal/pkg/testing/log"
+)
+
+func TestCheckCompatibilityInternal(t *testing.T) {
+	tests := []struct {
+		name         string
+		fleetVersion string
+		esVersion    string
+	}{
+		{
+			name:         "empty fleet and elasticsearch version",
+			fleetVersion: "",
+			esVersion:    "",
+		},
+		{
+			name:         "same version",
+			fleetVersion: "7.13.0",
+			esVersion:    "7.13.0",
+		},
+		{
+			name:         "new fleet-server patch",
+			fleetVersion: "7.13.2",
+			esVersion:    "7.13.1",
+		},
+		{
+			name:         "new es minor",
+			fleetVersion: "7.13.2",
+			esVersion:    "7.14.2",
+		},
+		{
+			name:         "new es major",
+			fleetVersion: "7.15.2",
+			esVersion:    "8.0.0",
+		},
+		{
+			name:         "new fleet-server minor",
+			fleetVersion: "7.14.0",
+			esVersion:    "7.13.1",
+		},
+		{
+			name:         "new fleet-server major",
+			fleetVersion: "8.0.0",
+			esVersion:    "7.18.0",
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := testlog.SetLogger(t).WithContext(t.Context())
+			err := checkCompatibility(ctx, tc.fleetVersion, tc.esVersion)
+			require.NoError(t, err, "expected snapshot version check should always succeed")
+		})
+	}
+}

--- a/internal/pkg/ver/es_check_test.go
+++ b/internal/pkg/ver/es_check_test.go
@@ -2,6 +2,8 @@
 // or more contributor license agreements. Licensed under the Elastic License;
 // you may not use this file except in compliance with the Elastic License.
 
+//go:build !snapshot
+
 package ver
 
 import (

